### ISSUE TITLE
Support for `numpy.ufunc.outer`

### ISSUE THREAD

**Issue by randolf-scholz** [google/jax#13054]
In `numpy`, universal functions support several augmented methods, among them:

- `ufunc.outer(A, B, /, **kwargs)` Apply the ufunc op to all pairs (a, b) with a in A and b in B.
- `ufunc.reduce(array[, axis, dtype, out, ...])` Reduces array's dimension by one, by applying ufunc along one axis.
- `ufunc.accumulate(array[, axis, dtype, out])` Accumulate the result of applying the operator to all elements.

See: https://numpy.org/doc/stable/reference/ufuncs.html#methods.

In particular, something like `add.outer(a,b)` come up surprisingly often.

**Comment by jakevdp** [comment]
Thanks - I have a draft PR that adds this functionality here: https://github.com/google/jax/pull/9529

We haven't gone farther wilth it because it adds a lot of complexity to the codebase and it's not clear that the benefit is commensurate, and it's not clear that JAX needs it.

In numpy, the benefit of having `outer` and other operations tied to ufuncs is that it provides a way to ensure that batched operations are lowered down to compiled loops. In JAX, we have the `vmap` transform that can do efficient batching for `any` function, not just ufuncs!

For example:
```python
import numpy as np
import jax.numpy as jnp
from jax import vmap

x = np.array([3, 4, 2])
y = np.array([5, 2, 3, 6])

print(np.multiply.outer(x, y))
# [[15  6  9 18]
#  [20  8 12 24]
#  [10  4  6 12]]
print(vmap(vmap(jnp.multiply, (None, 0)), (0, None))(x, y))
# [[15  6  9 18]
#  [20  8 12 24]
#  [10  4  6 12]]
```
In many ways, `vmap` subsumes the need to define an `outer` attribute on all individual functions, and the result is a simpler codebase. What do you think?

**Comment by JeppeKlitgaard** [comment]
I just ran into this as well and think it would be super neat to see `reduce` and other ufunc methods implemented in `jax`, which would ease changing over from `numpy`. 

Additionally I would argue that the `numpy` syntax is more readable than wrapping functions in layers of `vmap`s.

Alternatively maybe a helpful error could be added to a documentation page describing workarounds using `vmap`?

**Comment by jakevdp** [comment]
Thanks for the note - I should clean up and land my ufunc PR – I agree it would be nice to support this.

**Comment by phoenicyan** [comment]
also need the functionality... ETA?

**Comment by jakevdp** [comment]
Tracked in #18375

**Comment by billtubbs** [comment]
I would like to add my vote to this proposal.  Numpy's ufunc.accumulate method is useful for recurrent equations such as when simulating a dynamic system (i.e. recursively applying a state transition function to a state vector over multiple time-steps).

Simple example:
```python
import numpy as np

def lp_filter(x, u):
    return 0.3 * x + 0.7 * u

lp_filter_ufunc = np.frompyfunc(lp_filter, nin=2, nout=1)

U = np.zeros(11)
U[1:] = 1.
Y = lp_filter_ufunc.accumulate(U)
```

I guess you could do this with a for loop though.

What would be really nice, and as far as I know doesn't exist in Numpy, would be an accumulate method for general [vectorized](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.vectorize.html) functions that takes multi-dimensional inputs and has a multidimensional state.

**Comment by jakevdp** [comment]
We already have support for this – the general `accumulate` can be done via `jnp.frompyfunc`:
```python
import jax
import jax.numpy as jnp

def lp_filter(x, u):
    return 0.3 * x + 0.7 * u

lp_filter_ufunc = jnp.frompyfunc(lp_filter, nin=2, nout=1)

U = jnp.zeros(11).at[1:].set(1)
Y = lp_filter_ufunc.accumulate(U)
```
Accumulate here is implemented via `scan`.

For the vectorized accumulate, if I'm understanding what you have in mind, I think you can achieve that by wrapping the above function in `vmap`; for example:
```python
U = jnp.arange(15).reshape(3, 5)
print(jax.vmap(lp_filter_ufunc.accumulate)(U))
```
```
[[ 0.        0.7       1.61      2.583     3.5749  ]
 [ 5.        5.7       6.61      7.583     8.5749  ]
 [10.       10.7      11.61     12.582999 13.5749  ]]
```

**Comment by billtubbs** [comment]
Sorry.  I think it's because I'm using an old version of jax (0.4.14).  I got: `module 'jax.numpy' has no attribute 'frompyfunc'`.

Regarding the second part, I want to simulate a system where `u` and `x` are vectors.  For example:

```
A = jnp.array([
       [ -6. ,  -0. ,   0. ,  -5. ],
       [ -0. ,  -6. ,   0.5,  -0. ],
       [  0. , -10. ,   0. ,  -0. ],
       [  1. ,   0. ,   0. ,  -0. ]
])
B = jnp.array([
       [10.,  0.],
       [ 0.,  1.],
       [ 0.,  0.],
       [ 0.,  0.]
])

**Comment by billtubbs** [comment]
Actually, it can be done quite easily using Python's accumulate:
```python
from itertools import accumulate

x0 = jnp.zeros(4)
U = jnp.zeros((11, 2)).at[1:, :].set(1)
X = jnp.array(list(accumulate(U, f, initial=x0)))
```
Maybe I should just use this.  This will be differentiable right?

**Comment by jakevdp** [comment]
Ah, it sounds like you're asking for the JAX equivalent of this numpy feature request: https://github.com/numpy/numpy/issues/14020. I don't know of an easy API for this in JAX (you could always write a `scan` manually if you wish).

Regarding the `itertools` approach: it will work and be compatible with autodiff, but will become very slow as the size of the array becomes large, because this kind of python-side iteration is unrolled in JAX's compute graph.

**Comment by billtubbs** [comment]
Yes exactly.  I didn't know about that GUFunc proposal. That is awesome, I wish they would implement it.

I suspect you are right about the size of the computation graph in my application.  It may quickly become a limiting factor (I was planning to simulate for 60 time steps and x(k) might be length 20 or so).

There are better autodiff tools for simulating and optimizing dynamic systems (e.g. [CasADi](https://web.casadi.org)), which take account of the sparsity of the A, B matrices. The client of this project is familiar with Jax so I was trying to stay within their ecosystem.

Thanks for the feedback and sorry this was a bit off the topic of this thread.

**Comment by billtubbs** [comment]
Since you mentioned it, I checked out `scan` and it seems ideal for this. You can even include a mapping from states to outputs:

```
import jax
import jax.numpy as jnp

A = jnp.array([
       [ -6. ,  -0. ,   0. ,  -5. ],
       [ -0. ,  -6. ,   0.5,  -0. ],
       [  0. , -10. ,   0. ,  -0. ],
       [  1. ,   0. ,   0. ,  -0. ]
])
B = jnp.array([
       [10.,  0.],
       [ 0.,  1.],
       [ 0.,  0.],
       [ 0.,  0.]
])
C = jnp.array([
       [-0. , -0. , -0.5,  0.5],
       [ 1. , -0. , -1. ,  0.5]
])

def f(x, u):
    return A.dot(x) + B.dot(u)

def g(x, u):
    return C.dot(x)

def update(x, u):
    x = f(x, u)
    y = g(x, u)
    return x, y

x0 = jnp.zeros(4)
U = jnp.zeros((10, 2)).at[1:, :].set(1)
Y = jax.lax.scan(update, x0, U)
```
